lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -48,6 +48,182 @@ void _setStaticStyleAttributes(html.HtmlElement domElement) {
+          elements[autofill.uniqueIdentifier] = htmlElement;
+          formElement.append(htmlElement);
+        }
+      }
+    }
+
+    return AutofillGroup(
+      formElement: formElement,
+      elements: elements,
+      items: items,
+      singleElement: singleElement,
+    );
+  }
+
+  void placeForm(html.HtmlElement mainTextEditingElement) {
+    formElement.append(mainTextEditingElement);
+    domRenderer.glassPaneElement.append(formElement);
+  }
+
+  void removeForm() {
+    formElement.remove();
+  }
+
+  List<StreamSubscription<html.Event>> addEventListeners() {
+    Iterable<String> keys = elements.keys;
+    List<StreamSubscription<html.Event>> subscriptions =
+        List<StreamSubscription<html.Event>>();
+    keys.forEach((String key) {
+      final html.Element element = elements[key];
+      subscriptions.add(element.onInput.listen((html.Event e) {
+        _handleChange(e, element, key);
+      }));
+    });
+    return subscriptions;
+  }
+
+  void _handleChange(html.Event event, html.Element domElement, String tag) {
+    EditingState newEditingState = EditingState.fromDomElement(domElement);
+
+    assert(newEditingState != null);
+    _sendAutofillEditingState(tag, newEditingState);
+  }
+
+  /// Sends the 'TextInputClient.updateEditingStateWithTag' message to the framework.
+  void _sendAutofillEditingState(String tag, EditingState editingState) {
+    if (window._onPlatformMessage != null) {
+      window.invokeOnPlatformMessage(
+        'flutter/textinput',
+        const JSONMethodCodec().encodeMethodCall(
+          MethodCall(
+            'TextInputClient.updateEditingStateWithTag',
+            <dynamic>[
+              0,
+              <String, dynamic>{tag: editingState.toFlutter()}
+            ],
+          ),
+        ),
+        _emptyCallback,
+      );
+    }
+  }
+}
+
+/// Autofill related values.
+///
+/// These values are to be used when a text field have autofill enabled.
+@visibleForTesting
+class Autofill {
+  Autofill({this.editingState, this.uniqueIdentifier, this.hint});
+
+  final EditingState editingState;
+
+  final String uniqueIdentifier;
+
+  final String hint;
+
+  factory Autofill.fromFrameworkMessage(Map<String, dynamic> autofill) {
+    final String uniqueIdentifier = autofill['uniqueIdentifier'];
+    final List<dynamic> hintsList = autofill['hints'];
+    final EditingState editingState =
+        EditingState.fromFrameworkMessage(autofill['editingValue']);
+    return Autofill(
+        uniqueIdentifier: uniqueIdentifier,
+        hint: BrowserAutofillHints.instance.flutterToEngine(hintsList[0]),
+        editingState: editingState);
+  }
+
+  void applyToDomElement(html.HtmlElement domElement) {
+    domElement.id = hint;
+    if (domElement is html.InputElement) {
+      html.InputElement element = domElement;
+      element.name = hint;
+      element.id = hint;
+      element.autocomplete = hint;
+      element.type = 'text';
```

Comment:
This will override the type of the focused element (in some cases, the type is `password`).

Reply:
Good point! thanks! :)

Reply:
done